src/types/assertion.rs

```diff
@@ -0,0 +1,661 @@
+#[serde(rename_all = "snake_case")]
+pub struct Assertion {
+    root: Op,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(tag = "cmp")]
+#[serde(rename_all = "snake_case")]
+
+enum Comparison {
+    LessThan,
+    GreaterThan,
+    Equal,
+    NotEqual,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(tag = "header_op")]
+#[serde(rename_all = "snake_case")]
+enum HeaderOperand {
+    Literal { value: String },
+    Regex { value: String },
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(tag = "header_cmp")]
+#[serde(rename_all = "snake_case")]
+enum HeaderComparison {
+    Always,
+    Never,
+    Equals { test_value: HeaderOperand },
+    NotEquals { test_value: HeaderOperand },
+    LessThan { test_value: String },
+    GreaterThan { test_value: String },
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(tag = "op")]
+#[serde(rename_all = "snake_case")]
+enum Op {
+    And {
+        children: Vec<Op>,
+    },
+    Or {
+        children: Vec<Op>,
+    },
+    Not {
+        operand: Box<Op>,
+    },
+    StatusCodeCheck {
+        value: u16,
+        operator: Comparison,
+    },
+    // JsonPath {
+    //     value: JsonPathThingy,
+    // },
+    // XmlPath {
+    //     value: XmlPathThingy,
+    // },
+    HeaderCheck {
+        key: HeaderComparison,
+        value: HeaderComparison,
+    },
+}
+
+pub mod compiled {
+    use http::HeaderValue;
+    use regex::{Regex, RegexBuilder};
+    use std::str::FromStr;
+
+    // Don't make this any lower--there are some really innoccuous looking regexes (\d, \w) that
+    // actually expand to fairly large sizes, owing to large numbers of unicode characters.
+    const REGEX_SIZE_LIMIT: usize = 65536;
+
+    #[derive(thiserror::Error, Debug)]
+    pub enum Error {
+        #[error("Invalid regex syntax")]
+        InvalidRegex(String),
+
+        #[error("Regex is too complicated")]
+        RegexTooBig(usize),
+    }
+
+    #[derive(Debug, Clone, PartialEq, Eq)]
+    pub struct Assertion {
+        root: Op,
+    }
+
+    impl Assertion {
+        pub fn eval(
+            &self,
+            status_code: u16,
+            headers: &hyper::header::HeaderMap<HeaderValue>,
+            body: &str,
+        ) -> bool {
+            self.root.eval(status_code, headers, body)
+        }
+    }
+
+    #[derive(Debug, Clone, PartialEq, Eq)]
+    enum Comparison {
+        LessThan,
+        GreaterThan,
+        Equal,
+        NotEqual,
+    }
+
+    impl From<&super::Comparison> for Comparison {
+        fn from(value: &super::Comparison) -> Self {
+            match *value {
+                super::Comparison::LessThan => Comparison::LessThan,
+                super::Comparison::GreaterThan => Comparison::GreaterThan,
+                super::Comparison::Equal => Comparison::Equal,
+                super::Comparison::NotEqual => Comparison::NotEqual,
+            }
+        }
+    }
```

Comment:
What's the semantic difference between the `compiled::Comparison` and the `assertion::Comparison`?